### PR TITLE
Fix code that violated the (!Alive || !Reported) control-flow-analysis invariant.

### LINKED ISSUE #25043: Assertion failure in ControlFlowPass.IntersectWith with local function inside switch statement
1. Checkout master & run a debug build of VisualStudioSetup.Next
2. Paste the following code:
```c#
class C
{
    void M()
    {
        switch (true)
        {
            case true:
                void f()
                {
                }
        }
    }
}
```
3. Select the line with `switch (true)` or any line inside the switch statement. This repros most consistently when selecting the whole line by clicking in the left margin in VS.

This is what I get:
image

### PR DESCRIPTION
**Pull request by gafter**
### Customer scenario

An assertion error in the compiler arises when using a debug build of Roslyn and analyzing control-flow in code in which there is a "fall through" error in a switch statement.

### Bugs this fixes

Fixes #25043

### Workarounds, if any

None needed (customers do not usually use a Debug build).

### Risk

Tiny. The fix is the deletion of code that simply violates asserted invariants.

### Performance impact

Deletion of code could, if anything, speed up the compiler.

### Is this a regression from a previous update?

No; longstanding bug not previously detected.

### Root cause analysis

Unusual combination of conditions not hit in existing tests.

### How was the bug found?

Customer reported.

### Test documentation updated?

N/A

@dotnet/roslyn-compiler May I please have a couple of reviews of this ultra-tiny bug fix?

### CODE REVIEW

**Comment by gafter** [comment]
@agocke Could you please look at this super-tiny bug fix?

File: src/Compilers/CSharp/Portable/FlowAnalysis/ControlFlowPass.cs [hunk 1]
```diff
@@ -323,7 +323,6 @@ public override BoundNode VisitSwitchSection(BoundSwitchSection node, bool lastS
                 var boundLabel = node.SwitchLabels.Last();
                 Diagnostics.Add(lastSection ? ErrorCode.ERR_SwitchFallOut : ErrorCode.ERR_SwitchFallThrough,
                                 new SourceLocation(boundLabel.Syntax), boundLabel.Label.Name);
-                this.State.Reported = true;
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/FlowAnalysis/ControlFlowPass.cs]: The removal looks OK, but I'm trying to understand why this was here in the first place.

Was the intention that switch fall-through was considered unreachable code? I can see how it may be interpreted that way. And then the assert fired because the local function is still reachable?

Comment (gafter) [reply src/Compilers/CSharp/Portable/FlowAnalysis/ControlFlowPass.cs]: The idea of the `Reported` state was that once we reported an error for unreachable code (the principal error reported in the control flow pass), we don't need to report it on subsequent statements. We used the `Reported` bit to record the fact that the unreachable state was already reported as an error and therefore does not deserve another error. So the pattern was to set `Reported` whenever reporting the error.

But that pattern was followed too slavishly: we report an error here, but it is not an error of unreachable code (on the contrary, it is an error of *reachable* code where it should not be reachable). The assertion is that a state is either not reachable or not reported (because if it were reachable it should not have been reported). But setting that bit here clearly violates the invariant (see the enclosing `if`).

In short, the code being deleted here was simple a thinko.

Comment (agocke) [reply src/Compilers/CSharp/Portable/FlowAnalysis/ControlFlowPass.cs]: OK, so the state here is reachable. This matches my expectations.

**Comment by gafter** [comment]
@jaredpar Do I need approval to merge?

**Comment by jaredpar** [comment]
approved
